driver-core/src/main/com/mongodb/internal/connection/EventHelper.java

```diff
@@ -0,0 +1,128 @@
+            if (!equivalentEvent(curNew, matchingPrev)) {
+                return false;
+            }
+        }
+        return true;
+    }
+
+    static boolean equivalentEvent(final ServerDescription current, final ServerDescription previous) {
+        if (current == previous) {
+            return true;
+        }
+        if (previous == null || current == null) {
+            return false;
+        }
+        if (current.isOk() != previous.isOk()) {
+            return false;
+        }
+        if (current.getState() != previous.getState()) {
+            return false;
+        }
+        if (current.getType() != previous.getType()) {
+            return false;
+        }
+        if (current.getMinWireVersion() != previous.getMinWireVersion()) {
+            return false;
+        }
+        if (current.getMaxWireVersion() != previous.getMaxWireVersion()) {
+            return false;
+        }
+        if (!Objects.equals(current.getCanonicalAddress(), previous.getCanonicalAddress())) {
+            return false;
+        }
+        if (!current.getHosts().equals(previous.getHosts())) {
+            return false;
+        }
+        if (!current.getPassives().equals(previous.getPassives())) {
+            return false;
+        }
+        if (!current.getArbiters().equals(previous.getArbiters())) {
+            return false;
+        }
+        if (!current.getTagSet().equals(previous.getTagSet())) {
+            return false;
+        }
+        if (!Objects.equals(current.getSetName(), previous.getSetName())) {
+            return false;
+        }
+        if (!Objects.equals(current.getSetVersion(), previous.getSetVersion())) {
+            return false;
+        }
+        if (!Objects.equals(current.getElectionId(), previous.getElectionId())) {
+            return false;
+        }
+        if (!Objects.equals(current.getPrimary(), previous.getPrimary())) {
+            return false;
+        }
+        if (!Objects.equals(current.getLogicalSessionTimeoutMinutes(), previous.getLogicalSessionTimeoutMinutes())) {
+            return false;
+        }
+        if (!Objects.equals(current.getTopologyVersion(), previous.getTopologyVersion())) {
+            return false;
+        }
+        if (!exceptionsEquals(current.getException(), previous.getException())) {
+            return false;
+        }
+        return true;
+    }
+
+    @SuppressWarnings("BooleanMethodIsAlwaysInverted")
+    private static boolean exceptionsEquals(final Throwable current, final Throwable previous) {
+        // Compare class equality and message as exceptions rarely override equals
+        Class<?> thisExceptionClass = current != null ? current.getClass() : null;
+        Class<?> thatExceptionClass = previous != null ? previous.getClass() : null;
+        if (!Objects.equals(thisExceptionClass, thatExceptionClass)) {
+            return false;
+        }
+
+        String thisExceptionMessage = current != null ? current.getMessage() : null;
```

Comment:
Why did you decide to do your null checks this way instead of something like:
```
if (current == null || previous == null) {
      return current == previous
}
```
And then comparing the fields directly (which would save one null check on current and previous). Is it just best practice to always use null checks before accessing a value and then using Object.equals?

Reply:
I like that!  The reason I did it this way is because I had copied this code from ServerDescription.equals, which, because it was written at a time when we couldn't use Java 7 classes, didn't use Objects.equals.  Then when I realized I could use Objects.equals here I didn't think to simplify it further.

Reply:
Done. Please resolve if you like the change